User Profile: Initial beta overview documentations (#83766)
elastic/elasticsearch: x-pack/docs/en/security/authentication/user-profile.asciidoc

[role="xpack"]
[[user-profile]]
=== User profiles

beta::[]

User profiles provide persistent and stable representations of end-users.
Persistent means it exists without the need for the user to be online.
Stable means it has an unique identifier that does not change through
the lifetime of a deployment and can be used to refer to the associated
user.
Being able to uniquely refer to users is a critical function that underpins
important features like personalisation and collaboration in {kib}.

==== How user profile works

> In the {stack}, a user is required to authenticate each request. Each user is

Author: Thanks for rewriting this section. I adopted with a minor edit: Instead of saying a profile is associated with "one or more users". I changed to it to say "is associated to a user". I think it is confusing at this point to hint that profile support the same user from different realms. We do talk about it in the following section after taking about the basic. So I'll keep this section a bit simpler.

represented as an object that encapsulates its identity and privileges.

> This object is ephemeral especially for users authenticating from a realm

Author: I meant to use `object` for its vanilla meaning, that is, a thing. But I see how it can be interpreted as programming term in a doc like this.

backed by external identity providers, e.g. <<saml-realm,SAML>>.
There is simply no way to refer to a user without that user being online
at the moment.

User profiles solve the problem by creating profile documents for
authenticated users and persisting these documents in the profile
<<system-indices,system index>>. Each profile document is assigned a unique ID
which can be used to refer to the associated user.
The user profile essentially becomes the representation of the user,
even when it's not actively online.

When a user logs in to {kib}, {kib} creates a new profile or associates an
existing one for the user.
By using the unique ID of the user profile, {kib} can store user preferences
separately for each user, which is key to fine-grained level of customisation.
This unique ID also offers a way to uniquely refer to a user so that
{kib} can route messages and notifications to the appropriate user,
regardless of whether they're logged in.

> ===== Why a new unique ID instead of just the `username`?

Author: Good point. Thanks

The `username` is _not_ unique outside a single realm. Using just the `username`
is not possible to differentiate two users with the same `username` but from

> different realms.

Author: A separate unique ID also makes it somewhat easier to support `username` changes. But I leave it out since this possibility hasn't really been explored.

NOTE: For certain use cases where the same user can indeed authenticate with
different realms, consider using the <<security-domain,security domain>>
feature.

==== Create and manage user profiles

To create a new user profile or update an existing one, use the
<<security-api-activate-user-profile,activate user profile API>>. When you
submit a request, {es} attempts to locate an existing profile document for the
specified user. If one doesn't exist, {es} creates a new profile document.

In either case, the profile document captures the user's `full_name`, `email`,
`roles`, and `realms`, and also includes the profile unique ID, timestamp for

> the operation, and version control numbers. You can retrieve a user profile with

Author: I replaced it with `etc.`.

the <<security-api-get-user-profile,get user profile API>> by including the
profile's unique ID (`uid`).

On top of the user's basic information, you can add data to a profile document
with the <<security-api-update-user-profile-data,update user profile API>>. For
example, you can add user-specific preferences as part of the profile data.

The <<security-api-search-user-profile>> can be used to search profiles
that match given criteria. The search API is tailor-made for {kib}'s collaboration
feature and is not intended to support every conceivable piece of user-search
functionality.

==== Limitations

* Creating a new user profile requires a user's authentication details
(`username` and `password` or its
<<token-authentication-services,OAuth2 access token>>).
This means that a user must authenticate at least one time to create a

> profile document. If a user hasn't authenticated into the system,
> then it's not possible to refer to that user with user profile.

Reviewer: ```suggestion
profile document. Users who have never authenticated to {kib}
(or another profile aware application) will not have a user profile
and will not be returned by the 
<<security-api-search-user-profile,search user profile API>> 
```

[role="xpack"]
[[user-profile]]
=== User Profile

beta::[]

User Profile provides a way to uniquely and stably refer to a user.
Unique means each user has an identifier that is guaranteed to be unique within
a deployment. Stable means the identifier remains generally unchanged in the

> lifetime of a deployment.

Reviewer 2: I think this starts on the wrong footing. Are users not unique enough that we need profiles for it?

Author: I agree it's a bad openning statement because I didn't try to imply that users are not unique enough. In fact, it is quite the opposite, users (username + realm) are too unique. But we probably don't want these details in the beginning either. So I think the missing keyword here is "persistent" because users can be empheral but profiles are always persisted.

This may seem to be trivial but it underpins many other useful features.
For example, preferences can be customized and persisted separately for each user,
messages and notifications can be routed to the right user using its unique identifier.

The {stack} has the notion of "user" which is required to authenticate each request.
Internally, the user is represented as a `User` object that encapsulates its identity
and privileges. However, the `User` object itself is not sufficient to support referring

> to a user because:

Author: No, we don't. We should strive to convey the same/similar message without having to resort to the `User` object.

> to a user because:

Reviewer 3: I agree! I removed mention of the `User` object in 485baa5e7d02276f6003bd9d72d7098a7b1b28d6.

NOTE: For use cases where one individual can authenticate against
multiple realms, you can use the <<security-domain,security domain>> feature
so that these distinct users are considered to be the same identity
and share a single user profile.

==== Create and manage user profiles

To create a new user profile or update an existing one, use the
<<security-api-activate-user-profile,activate user profile API>>. When you
submit a request, {es} attempts to locate an existing profile document for the
specified user. If one doesn't exist, {es} creates a new profile document.

In either case, the profile document captures the user's `full_name`, `email`,
`roles`, and `realms`, and also includes the profile unique ID and timestamp of
the operation. You can retrieve a user profile with
the <<security-api-get-user-profile,get user profile API>> by including the
profile's unique ID (`uid`).

In addition to the user's basic information, you can add data to a profile document
with the <<security-api-update-user-profile-data,update user profile API>>. For
example, you can add user-specific preferences as part of the profile data.

> The <<security-api-suggest-user-profile,search user profile API>> can be used to search profiles

Reviewer 3: ```suggestion
Use the <<security-api-suggest-user-profile,search user profile API>> to search profiles
```

Stable means it has an unique identifier that does not change through
the lifetime of a deployment and can be used to refer to the associated
user.
Being able to uniquely refer to users regardless whether they are actively
online is a critical function that underpins important features like
personalisation and collaboration in {kib}.

==== How a user profile works

A user profile is the persistent record that the {stack} stores for each interactive user
that authenticates to {kib}.
This persistent record allows the {stack} to store user-level data (such as format and
notification preferences) for each user, and provide a unique identifier for the user record.

Because the {stack} has support for externally managed users (such as users who authenticate
via SAML, or users stored in an LDAP directory), we make a distinction between the _user_ and their
_profile_.

The term "users" refer to the entities that authenticate requests to the {stack}.
Each user has a username and a set of privileges (represented by <<roles,roles>>) that
determine which types of requests may be performed by that user.
This _user_ may be ephemeral - it could exist only for the duration of a request to

> an {es} API or for the lifetime of a session in {kib}. It cannot be retrieved once
> the session is complete, and does not support storage of preferences.

Reviewer: ```suggestion
an {es} API or for the lifetime of a session in {kib}. These ephemeral users 
cannot be retrieved once the session is complete, and do not support storage
of preferences.
```

Reviewer: I'm aware that I'm correcting my own text here :)

A "user profile" is the long term record associated with a _user_.
It has a unique identifier, it is searchable and it can be used to store personalisation
data and preferences for the user. It is a representation of the _user_ that persists

> and even when the _user_ itself has no active requests or sessions in the {stack}.

Reviewer: ```suggestion
even when the _user_ itself has no active requests or sessions in the {stack}.
```

[role="xpack"]
[[user-profile]]
=== User profiles

beta::[]

> 

Author: We briefly touched on this in a separate channel and agreed that this is the change we want. I have incorporated it in the latest commit. Thanks Adam!

User profiles provide persistent and stable representations of end-users.
Persistent means that a profile exists without the need for the user to be online.
Stable means it has an unique identifier that does not change through
the lifetime of a deployment and can be used to refer to the associated
user.
Being able to uniquely refer to users regardless whether they are actively
online is a critical function that underpins important features like
personalisation and collaboration in {kib}.

==== How a user profile works

> 

Author: Same as above
